### PR TITLE
Refactor configuration, logging and general structure

### PR DESCRIPTION
**Pull request by simt2**
As discussed yesterday with @josdotso and @rtluckie , I did some refactoring in mh. I think most of these changes are necessary for easy addition of new features and automated tests. I would like feedback to know if this is the right direction.

### Configuration changes
* Move global configuration in under the `mh` key in `MH_CONFIG`. This is the only breaking change in this PR to make it easier to understand how configuration is overridden later.
* Make it possible to override global configuration on a per-app basis. This lays the ground for features like native namespaces or even different kubectl contexts. 

Example configuration outlining this new structure:
```yaml
# should: seperate actual configuration("namespace") from information("maintainers")
mh:
  maintainers:
  - sitesar@cisco.com
  namespace: default
  printRendered: true
  targetContext: localhost
  team: sre

appSources:
- name: apps
  kind: configPath
  source: apps

apps:
- name: foo
- name: bar-0.1.0
  alias: bar
  key: .bar
  namespace: bar-namespace # bar is deployed in a different namespace than foo
```
An app definition in `MH_CONFIG` allows to override all global values, see the types `MHConfig` and `AppConfig` which is a superset of the first.
Global options could be moved to another key or to the topmost level again. TBD is also how to deal with attributes like `team` that are no configuration but information. Some could be kept for logging purposes, e.g. app-specific `maintainers`. 

### Changes in logging and general structure
* Reflect the above configuration logic entirely in yaml-tagged structs and merge them together where necessary
* Introduce a `NewApp` method to sanitize and initialize app configuration
* Code in `mhlib/` now only logs info messages
* Loggers should be passed around the code and extended with additional fields depending on the level they reach -> DRY logging 
* `mhlib/` does not exit on fatal errors anymore, but passes errors upstream to be handled by command logic
* Removed unnecessary and/or slow helper methods that were called sometimes multiple times per run 
* Check pointers for nil instead of values for zero 

I tested manually using [this](https://dpaste.cisco.com/HPOKt) configuration, not pasted here because it still includes internal information. WIP also because logging messages still need some love and more useful information.

### CODE REVIEW

File: mhlib/mhConfig.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+// DefaultMHConfig is the default mh config and will most likely be modified
+// during execution with the following sources by rising priority:
+//
+// 1. This default configuration
+// 2. The "mh" key in the configuration file
+// 3. Environment variables starting with "MH_"
+// 4. Command line flags
+// 5. app-specific overrides in MH_CONFIG.
```
Comment (simt2) [inline mhlib/mhConfig.go]: I'm not entirely happy with this order, moving 5. up between 2. and 3. might be a better solution. TBD

**Comment by josdotso** [comment]
> Global options could be moved to another key or to the topmost level again. TBD is also how to deal with attributes like team that are no configuration but information. Some could be kept for logging purposes, e.g. app-specific maintainers.

Maybe have a key under `mh:` called `tags` to handle the informational stuff.

**Comment by simt2** [comment]
> Maybe have a key under mh: called tags to handle the informational stuff.
Good idea. We could treat that as map of strings internally to enable arbitrary fields and get rid of specific informational in structs at all. Another idea: Let's call it `labels` and later translate it to k8s labels? It seems useful to me to label stuff created via MH and additionally add those labels.

**Comment by simt2** [comment]
* A mh config is now also represented by a struct.
* Added `AppFile` struct representing how to retrieve a mh app, this can later be extended with additional sources like git or s3. Also used to override the file of an app.
* Added `AppSourceConfig` struct representing AppSources actually configured in a mh config file and use it to create `AppSource` structs containing `AppFiles` during runtime. This avoids parsing all AppSources once per each App.

**Comment by simt2** [comment]
I have removed the `namespace` field for now and would like to implement that when this PR is merged later, it's big enough on it's own. Same goes for labels/tags.

File: mhlib/mhConfig.go (hunk 1 above)
Comment (drekle) [reply mhlib/mhConfig.go]: Your described order is what I would expect as well.

**Comment by simt2** [comment]
FYI: `configPath` AppSource not working as expected, I will debug that later today or tomorrow.

File: mhlib/appSource.go [hunk 2]
```diff
@@ -14,7 +14,79 @@
...
+// NewAppSource returns an AppSource based on a given AppSourceConfig and path to
+// mh configuration(for "configPath" kind of AppSources).
+//
+// Todo: Get rid of "configFile" argument in favor of defined kinds of
+// AppSources - see AppSourceConfig.
+func NewAppSource(config AppSourceConfig, configFile string) (*AppSource, error) {
+	// Build glob pattern depending on kind of AppSource
+	var pattern string
+	if config.Kind == "path" {
+		pattern = config.Source + "/*.y*ml"
```
Comment (drekle) [inline mhlib/appSource.go]: we use magic strings a lot.  Would prefer something like...
```
package main

import (
	"log"

	yaml "gopkg.in/yaml.v2"
)

type ConfigKind string

var Path ConfigKind = "path"
var ConfigPath ConfigKind = "configPath"

type AppSourceConfig struct {
	Kind   ConfigKind `yaml:"kind"`
	Name   string     `yaml:"name"`
	Source string     `yaml:"source"`
}

// AppSourceConfigs is an Array of AppSourceConfigs defined in a mh configuration file
type AppSourceConfigs []AppSourceConfig

func main() {
	example1 := `
kind: "path"
name: "name1"
source: "source1"
`
	example2 := `
kind: "configPath"
name: "name2"
source: "source2"
`
	var sc1 AppSourceConfig
	var sc2 AppSourceConfig
	yaml.Unmarshal([]byte(example1), &sc1)
	yaml.Unmarshal([]byte(example2), &sc2)
	for _, config := range []AppSourceConfig{sc1, sc2} {
		switch config.Kind {
		case Path:
			log.Printf("%s is path kind.", config.Name)
		case ConfigPath:
			log.Printf("%s is ConfigPath kind.", config.Name)
		}
	}
}
```
output...
```
API server listening at: 127.0.0.1:37022
2018/07/03 09:01:14 name1 is path kind.
2018/07/03 09:01:14 name2 is ConfigPath kind.
```

I see this string typing done a lot in the kubernetes API's.

Comment (simt2) [reply mhlib/appSource.go]: I agree on that we need a change for that. When writing this I was thing of duck-typing and ditching `kind` completely later in favor for pointer attributes like `configPath` and `path`, compare https://godoc.org/k8s.io/api/core/v1#PersistentVolumeSource .

**Comment by simt2** [comment]
@josdotso did you have time to take another look yet?

File: mhlib/mhConfig.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+var DefaultMHConfig = MHConfig{
+	Maintainers:   []string{"none"},
+	PrintRendered: false,
+	Simulate:      false,
+	TargetContext: "localhost",
+	Team:          "sopd-sre",
```
Comment (josdotso) [inline mhlib/mhConfig.go]: sopd-sre should be something more generic like `sre`

**Comment by josdotso** [comment]
@simt2 Is there a way we can make this backward compatible so that merge to master doesn't break the world?  Making mh config format a versioned artifact would be awesome.  This would follow the pattern set by k8s resources, docker-compose files, among other things.

**Comment by simt2** [comment]
@josdotso I'll add backwards compatibility and a deprecation notice, but only if we agree to move to a top-level `mh` key in the future. If unsure, I'll rather remove the key again. What's your take on that?

I agree on versioned configuration, but consider it out of scope for this PR.

**Comment by josdotso** [comment]
sgtm @simt2 .  The backwards compatibility just needs to exist for long enough to get the majority of our users moving toward the new format.  Even if the versioned format support comes online, I can foresee dropping support for v1 someday.

Thanks!

**Comment by simt2** [comment]
So the only attribute we actually need backwards compatibility because it would break things is `targetContext`, I added support for that.

**Comment by simt2** [comment]
@josdotso I don't seem to have write access in this repository, could you merge this PR for me, please?

**Comment by josdotso** [comment]
Awesome work. Thanks, @simt2 !

**Pull request merged**